core/commons/che-core-commons-lang/src/test/java/org/eclipse/che/commons/lang/TopologicalSortTest.java

```diff
@@ -0,0 +1,146 @@
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Stream;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+public class TopologicalSortTest {
+
+  @Test(dataProvider = "sortingCases")
+  public void shouldApplySorting(List<Var> list, List<Var> expectedSorting) throws Exception {
+    // when
+    sort(list);
+
+    // then
+    assertEquals(list, expectedSorting);
+  }
+
+  @DataProvider
+  public static Object[][] sortingCases() {
+    return new Object[][] {
+      // keep order if no dependencies
+      new Object[] {
+        vars(var("b", "."), var("a", "."), var("c", ".")),
+        vars(var("b", "."), var("a", "."), var("c", "."))
+      },
+
+      // sort dependents after dependencies
+      new Object[] {
+        vars(var("a", "b"), var("b", "."), var("c", ".")),
+        vars(var("b", "."), var("a", "b"), var("c", "."))
+      },
+
+      // sort dependents after dependencies, case 2
+      new Object[] {
+        vars(var("a", "bc"), var("b", "."), var("c", ".")),
+        vars(var("b", "."), var("c", "."), var("a", "bc"))
+      },
+
+      // sort dependents after dependencies, case 3
+      new Object[] {
+        vars(var("d", "."), var("a", "c"), var("b", "."), var("c", ".")),
+        vars(var("d", "."), var("b", "."), var("c", "."), var("a", "c"))
+      },
+
+      // test the robustness against cycles
+      new Object[] {vars(var("a", "a")), vars(var("a", "a"))},
+      new Object[] {vars(var("a", "b"), var("b", "a")), vars(var("a", "b"), var("b", "a"))},
+      new Object[] {
+        vars(var("a", "b"), var("b", "c"), var("c", "a")),
+        vars(var("a", "b"), var("b", "c"), var("c", "a"))
+      },
+      new Object[] {
+        vars(
+            var("f", "e"),
+            var("a", "bd"),
+            var("b", "c"),
+            var("e", "b"),
+            var("c", "a"),
+            var("d", ".")),
+        vars(
+            var("d", "."),
+            var("a", "bd"),
+            var("b", "c"),
+            var("c", "a"),
+            var("e", "b"),
+            var("f", "e"))
+      }
+    };
+  }
+
+  private static Var var(String name, String value) {
+    return new Var(name, value);
+  }
+
+  private static List<Var> vars(Var... vars) {
+    return Stream.of(vars).collect(toList());
+  }
+
+  private static void sort(List<Var> vars) {
+    Function<Var, String> idFunction = v -> v.name;
+    Function<Var, Set<String>> predecessors =
+        v ->
+            v.value
+                .chars()
+                .filter(Character::isAlphabetic)
+                .mapToObj(c -> String.valueOf((char) c))
+                .collect(toSet());
+
+    TopologicalSort<Var, String> sort = new TopologicalSort<>(idFunction, predecessors);
+    List<Var> newVars = sort.sort(vars);
+    vars.clear();
+    vars.addAll(newVars);
+  }
+
+  private static final class Var {
```

Comment:
It needs some time to understand that `name` is one character string, and value is multi-characters strings that actually contains dependency. Would not it be simpler if it would look like the following:
```java
  private static final class Var {

    private final Character name;
    private final Set<Character> dependencies;
    ...
}

```
WDYT?